Nuance cancel options for processors

mattwire opened the pull request:
Overview
----------------------------------------
Alternative to https://github.com/civicrm/civicrm-core/pull/17146

Before
----------------------------------------
https://github.com/civicrm/civicrm-core/pull/17146

After
----------------------------------------
Don't set `isNotifyProcessorOnCancelRecur` on propertyBag if it was not set by the user. This gives the payment processor effectively 3 states to choose from: user selected yes, user selected no, user did not make a selection.

In the "user did not make a selection" state it was probably because `supportsCancelRecurringNotifyOptional()` is set to FALSE and the processor can implement it's own logic in this case.

Technical Details
----------------------------------------


Comments
----------------------------------------
@eileenmcnaughton See 2nd commit - I've now implemented this on Stripe and think this small change is important.

eileenmcnaughton commented:
Ok that change makes sense  - my only concern is that your change is not 'protected' by either a code comment or a unit test. I think it's worth adding a code comment to make the importance of NULL vs bool clear to a reader of the code

eileenmcnaughton commented:
I just transferred the needs-documentation tag from the original PR

mattwire commented:
Documentation here: https://github.com/civicrm/civicrm-dev-docs/pull/812
I added a comment to `doCancelRecurring()` docblock.

Also changed `$propertyBag->getIsNotifyProcessorOnCancelRecur()` to `$propertyBag->getter('isNotifyProcessorOnCancelRecur', TRUE)` on `doCancelRecurring()` so we don't get an exception when unset.

eileenmcnaughton commented:
@artfulrobot can you check that last comment? I feel like that is not the recommended way to access 'could be null properties' but it gets into core like that it will become the standard way it winds up being done as this is one of our first 'models' of using property bag

eileenmcnaughton commented:
Let's switch to has + getX because that is the recommendation & this will be the source of copy & paste.

@artfulrobot can you post a recommended snippet (not sure if the above will be clear to @mattwire )

mattwire commented:
I used `->getter` as the more compact form of `->has && ->getX`. Have updated to use the longer form as this provides a reference implementation. Should be good now @artfulrobot @eileenmcnaughton
